keps/sig-scheduling/4815-dra-partitionable-devices/README.md

```diff
@@ -252,222 +252,470 @@ non-goals of this KEP.
-	ConsumesCapacityFrom []DeviceRef `json:"consumesCapacityFrom,omitempty"`
-
-	// Attributes defines the set of attributes for this device.
-	// The name of each attribute must be unique in that set.
-	//
-	// To ensure this uniqueness, attributes defined by the vendor
-	// must be listed without the driver name as domain prefix in
-	// their name. All others must be listed with their domain prefix.
-	//
-	// Conflicting attributes from those provided via mixins are
-	// overwritten by the ones provided here.
-	//
-	// The maximum number of attributes and capacities combined is 32.
-	//
-	// +optional
-	Attributes map[QualifiedName]DeviceAttribute `json:"attributes,omitempty"`
-
-	// Capacity defines the set of capacities for this device.
-	// The name of each capacity must be unique in that set.
-	//
-	// To ensure this uniqueness, capacities defined by the vendor
-	// must be listed without the driver name as domain prefix in
-	// their name. All others must be listed with their domain prefix.
-	//
-	// Conflicting capacities from those provided via mixins are
-	// overwritten by the ones provided here.
-	//
-	// The maximum number of attributes and capacities combined is 32.
-	//
-	// +optional
-	Capacity map[QualifiedName]DeviceCapacity `json:"capacity,omitempty"`
+  // Includes defines the set of device mixins that this device includes.
+  //
+  // The propertes of each included mixin are applied to this device in
+  // order. Conflicting properties from multiple mixins are taken from the
+  // last mixin listed that contains them. Properties set on the device will
+  // always override properties from mixins.
+  //
+  // The mixins referenced here must be defined in the same
+  // ResourceSlice.
+  //
+  // The maximum number of mixins that can be included is 8.
+  //
+  // +optional
+  // +listType=atomic
+  Includes []DeviceMixinRef
+
+  // Attributes defines the set of attributes for this device.
+  // The name of each attribute must be unique in that set.
+  //
+  // To ensure this uniqueness, attributes defined by the vendor
+  // must be listed without the driver name as domain prefix in
+  // their name. All others must be listed with their domain prefix.
+  //
+  // Conflicting attributes from those provided via mixins are
+  // overwritten by the ones provided here.
+  //
+  // The maximum number of attributes and capacities combined is 32.
+  //
+  // +optional
+  Attributes map[QualifiedName]DeviceAttribute
+
+  // Capacity defines the set of capacities for this mixin.
+  // The name of each capacity must be unique in that set.
+  //
+  // To ensure this uniqueness, capacities defined by the vendor
+  // must be listed without the driver name as domain prefix in
+  // their name. All others must be listed with their domain prefix.
+  //
+  // Conflicting capacities from those provided via other mixins are
+  // overwritten by the ones provided here.
+  //
+  // The maximum number of attributes and capacities combined is 32.
```

Comment:
I do worry 32 is too low. How many are on, say, a MIG partition today?

Reply:
So for the top-level partition (gpu-0) in the NVIDIA A100 example there are 12 attributes and 13 capacities. Note that this includes capacities like `memorySlice0...memorySlice7`. So a total of 25, which is kinda close to the max number, but also maybe not quite what it would look like for a real device.

This number is not changes in this PR, so maybe @klueska has some input here.

Reply:
@klueska had an "intrange" resource type concept in an early prototype. So rather than listing out memorySlice0..7, you would have a single intRange capacity [0..7], and your consumers would look like [0], [1], [0..1], etc.

This would be super useful to compact the inferentia example, too. And maybe even the TPU one if we could do, [0,1], [0,2] (ie, tuples not just ranges).

But parsing that as a resource.Quantity would be a major change since that's used all over.

Reply:
Ah, I was thinking that exposing the actual memory slices as capacities might not really be necessary? But I might be wrong here. What is the use-case for that?

Reply:
I don't think they need to be exposed to the user as capacities, but they are needed for the shared consumption for sure. I do know though that external controllers may want to "pin down" to a particular partition, if they don't want to let the scheduler pick the partition. I think other attributes probably service that purpose (for example, partition name), but I can imagine someone using the memorySliceX in a CEL expression for that purpose.